### PR TITLE
Split ReadCertificate functions into parsing files and bytes

### PR DESCRIPTION
**Pull request by dopey**

### CODE REVIEW

**Review by maraino — CHANGES_REQUESTED** [review]
We already have `ParseCertificate`, `ParseCertificateBundle`, and `ParseCertificateRequest` we don't need new methods that get bytes as the input.

File: pemutil/pem.go [hunk 1]
```diff
@@ -231,52 +241,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
-	if len(certs) == 0 {
-		return nil, errors.New("error parsing certificate: no certificate found")
+
+	// DER format (binary)
+	crt, err := x509.ParseCertificate(data)
+	if err != nil {
+		if ctx.filename == "" {
+			return nil, errors.Wrapf(err, "error parsing certificate bundle bytes input")
+		}
+		return nil, errors.Wrapf(err, "error parsing %s", ctx.filename)
```
Comment (maraino) [inline pemutil/pem.go:300]: Return an InvalidPEMError

File: pemutil/pem.go [hunk 2]
```diff
@@ -231,52 +241,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
-			continue
+// ParseCertificateRequest extracts the first *x509.CertificateRequest
+// from the given data.
+//
+// - supports PEM and DER certificate formats
+//   - DER format requires WithAllowDER() option
+//   - If a DER-formatted file is given only one certificate will be returned.
+func ParseCertificateRequest(data []byte, opts ...Options) (*x509.CertificateRequest, error) {
+	// Populate options
+	ctx := newContext("certificate-request-bytes")
```
Comment (maraino) [inline pemutil/pem.go:313]: ditto

File: pemutil/pem.go [hunk 3]
```diff
@@ -231,52 +241,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
+// ParseCertificateBundle returns a list of *x509.Certificate parsed from
+// the given bytes.
+//
+// - supports PEM and DER certificate formats
+//   - DER format requires WithAllowDER() option
+//   - If a DER-formatted file is given only one certificate will be returned.
+func ParseCertificateBundle(data []byte, opts ...Options) ([]*x509.Certificate, error) {
+	var err error
+	// Populate options
+	ctx := newContext("certificate-bundle-bytes")
```
[the remark this answers is not in the document]
Comment (maraino) [reply pemutil/pem.go:253]: In fact, we're not writing files, so a context can be just 
```go
var ctx context
if err := ctx.apply(opts); err != nil {
	return nil, err
}
```

File: pemutil/pem.go [hunk 4]
```diff
@@ -231,52 +241,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
-		if err != nil {
-			return nil, errors.Wrap(err, "error parsing certificate request")
+	if !ctx.allowDER {
+		return nil, &InvalidPEMError{
+			File:    ctx.filename,
+			Type:    PEMTypeCertificate,
+			Message: "input does not contain any PEM encoded blocks",
 		}
-		return csr, nil
 	}
```
Comment (maraino) [inline pemutil/pem.go:352]: ditto

File: pemutil/pem.go [hunk 5]
```diff
@@ -231,52 +241,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
 
-	return nil, errors.New("error parsing certificate request: no certificate found")
+	// DER format (binary)
+	csr, err := x509.ParseCertificateRequest(data)
+	if err != nil && ctx.filename == "" {
+		return nil, errors.Wrapf(err, "error parsing certificate request bytes input")
+	}
+	return csr, errors.Wrapf(err, "error parsing %s", ctx.filename)
```
Comment (maraino) [inline pemutil/pem.go:359]: ditto InvalidPEMError.

File: pemutil/pem.go [hunk 6]
```diff
@@ -355,83 +435,38 @@ func ReadCertificate(filename string, opts ...Options) (*x509.Certificate, error
...
-
-	// DER format (binary)
-	crt, err := x509.ParseCertificate(b)
-	if err != nil {
-		return nil, errors.Wrapf(err, "error parsing %s", filename)
-	}
-	return []*x509.Certificate{crt}, nil
+	return ParseCertificateBundle(b,
+		WithFilename(filename),
+		WithAllowDER(),
```
Comment (maraino) [inline pemutil/pem.go:451]:
```suggestion
```

File: pemutil/pem.go [hunk 7]
```diff
@@ -355,83 +435,38 @@ func ReadCertificate(filename string, opts ...Options) (*x509.Certificate, error
...
-			return csr, nil
-		}
-	}
-
-	// DER format (binary)
-	csr, err := x509.ParseCertificateRequest(b)
-	return csr, errors.Wrapf(err, "error parsing %s", filename)
+	return ParseCertificateRequest(b,
+		WithFilename(filename),
+		WithAllowDER(),
```
Comment (maraino) [inline pemutil/pem.go:468]:
```suggestion
```

**Review by maraino — CHANGES_REQUESTED** [review]

File: pemutil/pem.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dopey) [reply pemutil/pem.go:253]: context needs to be a pointer, so I'd have to instantiate it with `var ctx = &context{}`. Feel like it's nicer to do just do `newContext()`.

File: pemutil/pem.go [hunk 8]
```diff
@@ -231,52 +231,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
+		var block *pem.Block
+		var bundle []*x509.Certificate
+		for len(data) > 0 {
+			block, data = pem.Decode(data)
+			if block == nil {
+				break
+			}
+			if block.Type != "CERTIFICATE" {
+				continue
+			}
```
Comment (maraino) [inline pemutil/pem.go:264]: [CertPool.AppendCertsFromPEM](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/x509/cert_pool.go;l=226-228) has the condition that you have removed:
```go
if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
	continue
}
```

File: pemutil/pem.go [hunk 9]
```diff
@@ -231,52 +231,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
+			bundle = append(bundle, crt)
 		}
-		if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
-			continue
+		if len(bundle) == 0 {
+			return nil, &InvalidPEMError{
+				File:    ctx.filename,
+				Type:    PEMTypeCertificate,
+				Message: fmt.Sprintf("error parsing %s: does not contain a valid PEM encoded certificate", filename),
```
Comment (maraino) [inline pemutil/pem.go:280]: You will get a similar error by default.
```suggestion
```

File: pemutil/pem.go [hunk 10]
```diff
@@ -231,52 +231,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
+	// Populate options
+	ctx := newContext("")
+	if err := ctx.apply(opts); err != nil {
+		return nil, err
+	}
+
+	filename := "input"
+	if ctx.filename != "" {
+		filename = ctx.filename
+	}
```
Comment (maraino) [inline pemutil/pem.go:251]: I don't really like this.

File: pemutil/pem.go [hunk 11]
```diff
@@ -231,52 +231,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
 
-		cert, err := x509.ParseCertificate(block.Bytes)
-		if err != nil {
-			return nil, errors.Wrap(err, "error parsing certificate")
+	// DER format (binary)
+	crt, err := x509.ParseCertificate(data)
+	if err != nil {
+		return nil, &InvalidPEMError{
+			File:    ctx.filename,
+			Message: fmt.Sprintf("error parsing %s as DER format: %v", filename, err),
```
Comment (maraino) [inline pemutil/pem.go:291]: I think the errors are more verbose than necessary. 

A better approach might be to return simpler errors in the `Parse*` methods and wrap them in the `Read*` methods with the filename:
```go
return fmt.Errorf("error parsing %s: %w", filename, err)
```

Comment (maraino) [reply pemutil/pem.go:291]: Bellow you're doing something different:
```go
return nil, &InvalidPEMError{
	File: ctx.filename,
	Type: PEMTypeCertificateRequest,
	Err:  err,
```

File: pemutil/pem.go [hunk 12]
```diff
@@ -355,83 +428,36 @@ func ReadCertificate(filename string, opts ...Options) (*x509.Certificate, error
...
-
-	// DER format (binary)
-	crt, err := x509.ParseCertificate(b)
-	if err != nil {
-		return nil, errors.Wrapf(err, "error parsing %s", filename)
-	}
-	return []*x509.Certificate{crt}, nil
+	return ParseCertificateBundle(b,
+		WithFilename(filename),
+	)
```
Comment (maraino) [inline pemutil/pem.go:444]: We can get rid of the filename and options in the parse if we do:

```
certs, err := ParseCertificateBundle(b)
if err != nil {
	return fmt.Errorf("error parsing %s: %w", filename, err)
}
```

File: pemutil/pem.go [hunk 13]
```diff
@@ -355,83 +428,36 @@ func ReadCertificate(filename string, opts ...Options) (*x509.Certificate, error
...
-			return csr, nil
-		}
-	}
-
-	// DER format (binary)
-	csr, err := x509.ParseCertificateRequest(b)
-	return csr, errors.Wrapf(err, "error parsing %s", filename)
+	return ParseCertificateRequest(b,
+		WithFilename(filename),
+	)
```
Comment (maraino) [inline pemutil/pem.go:460]: Ditto

**Review by maraino — CHANGES_REQUESTED** [review]

File: pemutil/pem.go [hunk 14]
```diff
@@ -231,52 +231,93 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
...
 		}
-		return csr, nil
 	}
 
-	return nil, errors.New("error parsing certificate request: no certificate found")
+	// DER format (binary)
+	csr, err := x509.ParseCertificateRequest(data)
+	if err != nil {
+		return nil, &InvalidPEMError{
+			Message: fmt.Sprint("error parsing certificate request as DER format", err),
```
Comment (maraino) [inline pemutil/pem.go:316]:
```suggestion
			Message: fmt.Sprint("error parsing certificate request as DER format: %v", err),
```

**Review by maraino — CHANGES_REQUESTED** [review]

**Review by maraino — APPROVED** [review]

**Pull request merged**